lxd/auth/driver_tls.go

```diff
@@ -68,11 +68,15 @@ func (t *tls) CheckPermission(ctx context.Context, r *http.Request, entityURL *a
 		return api.StatusErrorf(http.StatusForbidden, "Certificate is restricted")
 	}
 
-	entityType, projectName, _, _, err := entity.ParseURL(entityURL.URL)
+	entityType, projectName, _, pathArgs, err := entity.ParseURL(entityURL.URL)
 	if err != nil {
 		return fmt.Errorf("Failed to parse entity URL: %w", err)
 	}
 
+	if entityType == entity.TypeProject {
```

Comment:
Just an observation, but given the function is called `ParseURL` and returns a variable called projectName, should it not be the job of the `ParseURL` function to have this path specific logic in it and return projectName correctly if the URL is for a project?

Reply:
Also, and this isn't for now! But I as I am growing more familiar with the shared/entity package I am thinking  we should refactor the entity types to be just that, their own Go type (rather than a string constant as they are today), with properties & functions.

Then functions like `requiresProject` and `path` would just call the associated function on the specific entity type.

This way it will keep all entity type properties together in a single entity specific type, rather than spreading them over several different functions. This would seem to me to be a more maintainable solution as one would be prompted to think about these decisions for each new entity type being added in the future, rather than having to remember "ah, i need to also update `requiresProject` and `path`" for example.

Please could you open an issue (or add a task to an existing issue) to review this to make it more maintenance friendly in the future. Ta

Reply:
Yes, the name of the returned variable should really be `projectQueryParameter`, similarly `location` should be `targetQueryParameter`. 

The key thing that I want to maintain about `entity.ParseURL` and `(entity.Type).URL` is that they are each other's inverse. If the entity type is `project` and it returns a project name query parameter, the inverse will give us `/1.0/projects/my-project?project=my-project`. Maybe we can improve the logic there though.

Reply:
Yeah we could have the inverse logic in `(entity.Type).URL` to not append project= for a project. 
Feels cleaner and less error prone to me.

Reply:
https://github.com/canonical/lxd/issues/12928